D1: 'wrangler d1 migrations apply --remote' fails with 'incomplete input: SQLITE_ERROR' on CRLF migration files containing triggers

### Which Cloudflare product(s) does this pertain to?

D1, Wrangler

### What versions are you using?

Reproduced on wrangler 4.107.0 and 4.118.0 (Windows 11, Node 24).

### What operating system and version are you using?

Windows 11 Pro

### Describe the Bug

`wrangler d1 migrations apply <db> --remote` fails with:

```
X [ERROR] A request to the Cloudflare API (.../d1/database/<uuid>/query) failed.
  incomplete input: SQLITE_ERROR [code: 7500]
```

whenever a migration file has **CRLF line endings** and contains a **trigger** (`BEGIN ... END;` body). This is the default situation for Windows checkouts with `core.autocrlf=true`.

Minimal reproduction (5 lines). Save with **CRLF** line endings as `migrations/0001_repro.sql`:

```sql
CREATE TABLE probe_z (id TEXT PRIMARY KEY);
CREATE TRIGGER trg_z BEFORE DELETE ON probe_z
BEGIN
  SELECT RAISE(ABORT, 'no');
END;
```

Then: `wrangler d1 migrations apply <db> --remote` → fails as above.

Isolation matrix (each case verified on a fresh throwaway remote D1):

| Case | Result |
|---|---|
| The file above with **CRLF** endings via `d1 migrations apply --remote` | ❌ `incomplete input: SQLITE_ERROR [7500]` |
| The **identical bytes** via `d1 execute --remote --file` | ✅ |
| The same content with **LF** endings via `d1 migrations apply --remote` | ✅ |
| A CRLF migration **without** a trigger via `d1 migrations apply --remote` | ✅ |
| The same CRLF+trigger file via `d1 migrations apply --local` | ✅ |

Notes:

- Failures are **atomic**: the failed migration is not recorded in `d1_migrations` and no partial schema objects are created; earlier migrations in the same run apply and are recorded normally.
- Because `d1 execute --remote --file` succeeds with the identical bytes, the fault appears specific to the migrations-apply path's statement handling (client-side preprocessing or that path's transport of statements), rather than SQLite itself.
- Larger real-world migrations show the same signature and can appear content-dependent (comments/literals), but controlled byte-exact tests show line endings are the load-bearing variable.

Workaround: force LF for migration files via `.gitattributes` (`migrations/*.sql text eol=lf`).

### Please provide a link to a minimal reproduction

The 5-line SQL file above is the complete reproduction; it contains no application data.

--- a/.changeset/d1-crlf-remote-query.md
+++ b/.changeset/d1-crlf-remote-query.md
@@ -2,6 +2,6 @@
 "wrangler": patch
 ---
 
-Normalize CRLF line endings before sending D1 commands to the remote query API
+Normalize structural CRLF line endings before sending D1 commands to the remote query API
 
-`wrangler d1 migrations apply --remote` and `wrangler d1 execute --remote --command` failed with `incomplete input: SQLITE_ERROR` when the SQL contained CRLF line endings inside a compound statement such as a `CREATE TRIGGER ... BEGIN ... END;` body. The command string is now normalized to LF before it is sent to the D1 query API.
+`wrangler d1 migrations apply --remote` and `wrangler d1 execute --remote --command` failed with `incomplete input: SQLITE_ERROR` when the SQL contained CRLF line endings inside a compound statement such as a `CREATE TRIGGER ... BEGIN ... END;` body. Structural line endings are now normalized to LF before the command is sent to the D1 query API, while CRLF inside quoted values and identifiers remains unchanged.

--- a/.changeset/d1-crlf-remote-query.md
+++ b/.changeset/d1-crlf-remote-query.md
@@ -2,6 +2,6 @@
 "wrangler": patch
 ---
 
-fix: normalize CRLF line endings before sending D1 commands to the remote query API
+Normalize CRLF line endings before sending D1 commands to the remote query API
 
 `wrangler d1 migrations apply --remote` and `wrangler d1 execute --remote --command` failed with `incomplete input: SQLITE_ERROR` when the SQL contained CRLF line endings inside a compound statement such as a `CREATE TRIGGER ... BEGIN ... END;` body. The command string is now normalized to LF before it is sent to the D1 query API.

--- a/.changeset/d1-crlf-remote-query.md
+++ b/.changeset/d1-crlf-remote-query.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+fix: normalize CRLF line endings before sending D1 commands to the remote query API
+
+`wrangler d1 migrations apply --remote` and `wrangler d1 execute --remote --command` failed with `incomplete input: SQLITE_ERROR` when the SQL contained CRLF line endings inside a compound statement such as a `CREATE TRIGGER ... BEGIN ... END;` body. The command string is now normalized to LF before it is sent to the D1 query API.

--- a/packages/wrangler/src/__tests__/d1/execute.test.ts
+++ b/packages/wrangler/src/__tests__/d1/execute.test.ts
@@ -302,9 +302,62 @@ To continue without logging in, rerun this command with \`--temporary\`. Wrangle
 			expect(std.out).toMatch("🚣 Executed 1 command in 123.46ms");
 		});
 
-		it("should format batch execution duration with 2 decimal places", async ({
-			expect,
-		}) => {
+	it("should normalize CRLF line endings in commands sent to the remote query API", async ({
+		expect,
+	}) => {
+		setIsTTY(false);
+		writeWranglerConfig({
+			d1_databases: [
+				{ binding: "DATABASE", database_name: "db", database_id: "xxxx" },
+			],
+		});
+
+		msw.use(
+			...getMswSuccessMembershipHandlers([
+				{
+					id: "some-account-id",
+					name: "test-account",
+				},
+			])
+		);
+
+		let sentSql: string | undefined;
+		msw.use(
+			http.get("*/accounts/:accountId/d1/database", async () => {
+				return HttpResponse.json(
+					createFetchResult([
+						{ uuid: "xxxx", name: "db", created_at: "", version: "alpha" },
+					])
+				);
+			}),
+			http.post(
+				"*/accounts/:accountId/d1/database/:databaseId/query",
+				async ({ request }) => {
+					sentSql = ((await request.json()) as { sql: string }).sql;
+					return HttpResponse.json(
+						createFetchResult([
+							{
+								results: [{ result: 1 }],
+								success: true,
+								meta: { duration: 100 },
+							},
+						])
+					);
+				}
+			)
+		);
+
+		await runWrangler(
+			"d1 execute db --remote --command 'CREATE TRIGGER trg BEFORE DELETE ON probe_z\r\nBEGIN\r\n  SELECT RAISE(ABORT, '\''no'\'');\r\nEND;'"
+		);
+
+		expect(sentSql).toBeDefined();
+		expect(sentSql).not.toContain("\r");
+	});
+
+	it("should format batch execution duration with 2 decimal places", async ({
+		expect,
+	}) => {
 			setIsTTY(false);
 			writeWranglerConfig({
 				d1_databases: [

--- a/packages/wrangler/src/__tests__/d1/execute.test.ts
+++ b/packages/wrangler/src/__tests__/d1/execute.test.ts
@@ -302,7 +302,7 @@ To continue without logging in, rerun this command with \`--temporary\`. Wrangle
 			expect(std.out).toMatch("🚣 Executed 1 command in 123.46ms");
 		});
 
-		it("should normalize CRLF line endings in commands sent to the remote query API", async ({
+		it("should preserve quoted CRLF when normalizing commands sent to the remote query API", async ({
 			expect,
 		}) => {
 			setIsTTY(false);
@@ -348,11 +348,12 @@ To continue without logging in, rerun this command with \`--temporary\`. Wrangle
 			);
 
 			await runWrangler(
-				"d1 execute db --remote --command 'CREATE TRIGGER trg BEFORE DELETE ON probe_z\r\nBEGIN\r\n  SELECT RAISE(ABORT, '\''no'\'');\r\nEND;'"
+				"d1 execute db --remote --command \"CREATE TRIGGER trg BEFORE DELETE ON probe_z\r\nBEGIN\r\n  SELECT RAISE(ABORT, 'no\r\nchange');\r\nEND;\""
 			);
 
-			expect(sentSql).toBeDefined();
-			expect(sentSql).not.toContain("\r");
+			expect(sentSql).toBe(
+				"CREATE TRIGGER trg BEFORE DELETE ON probe_z\nBEGIN\n  SELECT RAISE(ABORT, 'no\r\nchange');\nEND;"
+			);
 		});
 
 		it("should format batch execution duration with 2 decimal places", async ({

--- a/packages/wrangler/src/__tests__/d1/execute.test.ts
+++ b/packages/wrangler/src/__tests__/d1/execute.test.ts
@@ -302,62 +302,62 @@ To continue without logging in, rerun this command with \`--temporary\`. Wrangle
 			expect(std.out).toMatch("🚣 Executed 1 command in 123.46ms");
 		});
 
-	it("should normalize CRLF line endings in commands sent to the remote query API", async ({
-		expect,
-	}) => {
-		setIsTTY(false);
-		writeWranglerConfig({
-			d1_databases: [
-				{ binding: "DATABASE", database_name: "db", database_id: "xxxx" },
-			],
-		});
+		it("should normalize CRLF line endings in commands sent to the remote query API", async ({
+			expect,
+		}) => {
+			setIsTTY(false);
+			writeWranglerConfig({
+				d1_databases: [
+					{ binding: "DATABASE", database_name: "db", database_id: "xxxx" },
+				],
+			});
 
-		msw.use(
-			...getMswSuccessMembershipHandlers([
-				{
-					id: "some-account-id",
-					name: "test-account",
-				},
-			])
-		);
+			msw.use(
+				...getMswSuccessMembershipHandlers([
+					{
+						id: "some-account-id",
+						name: "test-account",
+					},
+				])
+			);
 
-		let sentSql: string | undefined;
-		msw.use(
-			http.get("*/accounts/:accountId/d1/database", async () => {
-				return HttpResponse.json(
-					createFetchResult([
-						{ uuid: "xxxx", name: "db", created_at: "", version: "alpha" },
-					])
-				);
-			}),
-			http.post(
-				"*/accounts/:accountId/d1/database/:databaseId/query",
-				async ({ request }) => {
-					sentSql = ((await request.json()) as { sql: string }).sql;
+			let sentSql: string | undefined;
+			msw.use(
+				http.get("*/accounts/:accountId/d1/database", async () => {
 					return HttpResponse.json(
 						createFetchResult([
-							{
-								results: [{ result: 1 }],
-								success: true,
-								meta: { duration: 100 },
-							},
+							{ uuid: "xxxx", name: "db", created_at: "", version: "alpha" },
 						])
 					);
-				}
-			)
-		);
+				}),
+				http.post(
+					"*/accounts/:accountId/d1/database/:databaseId/query",
+					async ({ request }) => {
+						sentSql = ((await request.json()) as { sql: string }).sql;
+						return HttpResponse.json(
+							createFetchResult([
+								{
+									results: [{ result: 1 }],
+									success: true,
+									meta: { duration: 100 },
+								},
+							])
+						);
+					}
+				)
+			);
 
-		await runWrangler(
-			"d1 execute db --remote --command 'CREATE TRIGGER trg BEFORE DELETE ON probe_z\r\nBEGIN\r\n  SELECT RAISE(ABORT, '\''no'\'');\r\nEND;'"
-		);
+			await runWrangler(
+				"d1 execute db --remote --command 'CREATE TRIGGER trg BEFORE DELETE ON probe_z\r\nBEGIN\r\n  SELECT RAISE(ABORT, '\''no'\'');\r\nEND;'"
+			);
 
-		expect(sentSql).toBeDefined();
-		expect(sentSql).not.toContain("\r");
-	});
+			expect(sentSql).toBeDefined();
+			expect(sentSql).not.toContain("\r");
+		});
 
-	it("should format batch execution duration with 2 decimal places", async ({
-		expect,
-	}) => {
+		it("should format batch execution duration with 2 decimal places", async ({
+			expect,
+		}) => {
 			setIsTTY(false);
 			writeWranglerConfig({
 				d1_databases: [

--- a/packages/wrangler/src/__tests__/d1/splitter.test.ts
+++ b/packages/wrangler/src/__tests__/d1/splitter.test.ts
@@ -1,5 +1,22 @@
 import { describe, it } from "vitest";
-import { mayContainMultipleStatements, splitSqlQuery } from "../../d1/splitter";
+import {
+	mayContainMultipleStatements,
+	normalizeSqlLineEndings,
+	splitSqlQuery,
+} from "../../d1/splitter";
+
+describe("normalizeSqlLineEndings()", () => {
+	it("should preserve CRLF inside quoted SQL values and identifiers", ({
+		expect,
+	}) => {
+		const sql =
+			"SELECT 'single''quote\r\nvalue', \"double\r\nquote\", `backtick\r\nquote`, [bracket\r\nquote]; -- don't stop scanning\r\n/* block\r\ncomment */\r\nSELECT 1;";
+
+		expect(normalizeSqlLineEndings(sql)).toBe(
+			"SELECT 'single''quote\r\nvalue', \"double\r\nquote\", `backtick\r\nquote`, [bracket\r\nquote]; -- don't stop scanning\n/* block\ncomment */\nSELECT 1;"
+		);
+	});
+});
 
 describe("mayContainMultipleStatements()", () => {
 	it("should return false if there is only a semi-colon at the end", ({

--- a/packages/wrangler/src/d1/execute.ts
+++ b/packages/wrangler/src/d1/execute.ts
@@ -505,14 +505,17 @@ async function executeRemotely({
 			},
 		];
 	} else {
+		// The D1 query API splits multi-statement SQL on `;` server-side, and
+		// mishandles CRLF line endings inside compound statements such as a
+		// `CREATE TRIGGER ... BEGIN ... END;` body (issue #14991). Normalize
+		// to LF so the server receives the same input that works for LF files.
+		const sql = input.command?.replace(/\r\n/g, "\n");
 		const result = await d1ApiPost<QueryResult[]>(
 			config,
 			accountId,
 			db,
 			"query",
-			{
-				sql: input.command,
-			}
+			{ sql }
 		);
 		logResult(result);
 		return result;

--- a/packages/wrangler/src/d1/execute.ts
+++ b/packages/wrangler/src/d1/execute.ts
@@ -21,7 +21,7 @@ import { confirm } from "../dialogs";
 import { logger } from "../logger";
 import { readableRelative } from "../paths";
 import { requireAuth } from "../user";
-import { splitSqlQuery } from "./splitter";
+import { normalizeSqlLineEndings, splitSqlQuery } from "./splitter";
 import { getDatabaseByNameOrBinding, getDatabaseInfoFromConfig } from "./utils";
 import type {
 	Database,
@@ -508,8 +508,8 @@ async function executeRemotely({
 		// The D1 query API splits multi-statement SQL on `;` server-side, and
 		// mishandles CRLF line endings inside compound statements such as a
 		// `CREATE TRIGGER ... BEGIN ... END;` body (issue #14991). Normalize
-		// to LF so the server receives the same input that works for LF files.
-		const sql = input.command?.replace(/\r\n/g, "\n");
+		// structural line endings while preserving quoted SQL values.
+		const sql = input.command && normalizeSqlLineEndings(input.command);
 		const result = await d1ApiPost<QueryResult[]>(
 			config,
 			accountId,

--- a/packages/wrangler/src/d1/splitter.ts
+++ b/packages/wrangler/src/d1/splitter.ts
@@ -37,6 +37,98 @@ export function splitSqlQuery(sql: string): string[] {
 	}
 }
 
+/**
+ * Normalize structural CRLF line endings without changing quoted SQL values or
+ * identifiers.
+ */
+export function normalizeSqlLineEndings(sql: string): string {
+	let normalized = "";
+	let quoteEnd: "'" | '"' | "`" | "]" | undefined;
+	let inLineComment = false;
+	let inBlockComment = false;
+
+	for (let index = 0; index < sql.length; index++) {
+		const char = sql[index];
+		const nextChar = sql[index + 1];
+
+		if (quoteEnd !== undefined) {
+			normalized += char;
+			if (char === quoteEnd) {
+				if (nextChar === quoteEnd) {
+					normalized += nextChar;
+					index++;
+				} else {
+					quoteEnd = undefined;
+				}
+			}
+			continue;
+		}
+
+		if (inLineComment) {
+			if (char === "\r" && nextChar === "\n") {
+				normalized += "\n";
+				index++;
+				inLineComment = false;
+			} else {
+				normalized += char;
+				inLineComment = char !== "\n";
+			}
+			continue;
+		}
+
+		if (inBlockComment) {
+			if (char === "\r" && nextChar === "\n") {
+				normalized += "\n";
+				index++;
+			} else {
+				normalized += char;
+				if (char === "*" && nextChar === "/") {
+					normalized += nextChar;
+					index++;
+					inBlockComment = false;
+				}
+			}
+			continue;
+		}
+
+		if (char === "-" && nextChar === "-") {
+			normalized += "--";
+			index++;
+			inLineComment = true;
+			continue;
+		}
+
+		if (char === "/" && nextChar === "*") {
+			normalized += "/*";
+			index++;
+			inBlockComment = true;
+			continue;
+		}
+
+		if (char === "'" || char === '"' || char === "`") {
+			normalized += char;
+			quoteEnd = char;
+			continue;
+		}
+
+		if (char === "[") {
+			normalized += char;
+			quoteEnd = "]";
+			continue;
+		}
+
+		if (char === "\r" && nextChar === "\n") {
+			normalized += "\n";
+			index++;
+			continue;
+		}
+
+		normalized += char;
+	}
+
+	return normalized;
+}
+
 function splitSqlIntoStatements(sql: string): string[] {
 	const statements: string[] = [];
 	let str = "";

